### ISSUE TITLE
The pain of maintaining a polymer app

### ISSUE THREAD

**Issue by robdodson** [PolymerElements/polymer-starter-kit#858]
_Moved from https://github.com/google/WebFundamentals/issues/3044_

Some things I found while building on starter kit

* bower with its "questions" really raises the question if components still work. 
* the community is not so fast. Maybe a little more tooling would help. 
* starter kit takes all the elements, maybe introduce users to a little more choices. Selecting and installing just a part of the components is not so common. The general polymer world has some x megabyte repos in terms of dependencies for one element. Is that necessary, or just a result of development docs saying so?  
* starter kit installs the test toolchain multiple times. that takes time 
* To upgrade a old application the best way seems to reinstall all modules via bower. Using the offcial bower repo was my choice! Not using the direct git urls anymore. 

I came from pre 1.0 and now is 1.6.0. Updating takes a lot of effort. I know a lot of these issues are with 3rd party modules. But maybe have a closer look at some elements that are reularly used and re-used and maybe support these a little more. Maybe a kind of validator and wizard can help generate the right pull requests for these. Online and connected to github. 

In a nutshell. 

Please, 

* make starter kit a little leaner and teach users to build small
* help upgrades with even more tooling. Non Module authors must be able to shortcircuit a dead component a little better. 
* try reduce the pain form quirky bower installs please 

cc @sebs

**Comment by robdodson** [comment]
> bower with its "questions" really raises the question if components still work.

Can you explain what you mean by "questions"? Are you referring to when bower asks you to resolve a version conflict?

> the community is not so fast. Maybe a little more tooling would help.

Not sure what you mean by this? What do you want the community to be "faster" at? Like releasing more elements or...?

> starter kit takes all the elements, maybe introduce users to a little more choices. Selecting and installing just a part of the components is not so common. The general polymer world has some x megabyte repos in terms of dependencies for one element. Is that necessary, or just a result of development docs saying so?

I'm not sure I understand the question but typically we install all of the elements in Starter Kit because it's easier to install everything and have it local on disk versus having to install a bunch of individual elements. Ultimately the only elements that are actually shipped in production are the ones you end up using in your app. For example: we install all `paper-elements` but if you only import `paper-input` in your app then when you vulcanize only `paper-input` will be shipped to the client.

> starter kit installs the test toolchain multiple times. that takes time

This is an interesting point, let me get some clarification on it. @azakus we have web-component-tester in both our bower.json and package.json. Do we need it in both places?

> To upgrade a old application the best way seems to reinstall all modules via bower. Using the offcial bower repo was my choice! Not using the direct git urls anymore.

I don't understand what you mean by "using the official bower repo was my choice!". The official bower repo of what?

> I came from pre 1.0 and now is 1.6.0. Updating takes a lot of effort. I know a lot of these issues are with 3rd party modules.

We don't have much control here. It would be an incredible amount of work to try to auto-pull request 3rd party modules. It's ultimately up to the authors to keep pace with our releases. Going from pre 1.0 to 1.6 is indeed a massive leap (literally 1+ years worth of versions) so it's not surprising that there was breakage. Post 1.0 things should be much more stable.

> make starter kit a little leaner and teach users to build small

[This is what psk2 is doing](https://github.com/PolymerElements/polymer-starter-kit/tree/psk2) and the reason we built the CLI so it can stamp out basic starting points. For example the app-drawer-template that comes with CLI is very lean.

> try reduce the pain form quirky bower installs please

we would like to move off of bower and onto npm. I agree it's annoying having to do manual version conflict resolution and we plan to explore how to make this a bit easier.

**Comment by sebs** [comment]
Questions: bower deps, yeah. 

Community not so fast: When a version jumps comes, push the community to release updated elements. A missing "data-table" was my killer in the last one. Another one a paper-date-select. Both seem to be more "needed in teh core" than 3rd party.

> I don't understand what you mean by "using the official bower repo was my choice!". The official bower repo of what?

Bower deps seem to point on github repos, I point them just on the name now. 

> We don't have much control here. It would be an incredible amount of work to try to auto-pull request 3rd party modules. It's ultimately up to the authors to keep pace with our releases. Going from pre 1.0 to 1.6 is indeed a massive leap (literally 1+ years worth of versions) so it's not surprising that there was breakage. Post 1.0 things should be much more stable.

Yeah, just had teh feeling of "I already did this" with the move to 1.x (I guess) and now again 

> we would like to move off of bower and onto npm. I agree it's annoying having to do manual version conflict resolution and we plan to explore how to make this a bit easier.

The only reason not to  write a short "just move to npm" was sort of my concern that you would not think of it for what ever reason I imagined. Would be on board to help port stuff even. It would really fix a lot of problems ;)))) And give some extra speed. 

Thanks for the reply ;)

**Comment by sebs** [comment]
Ah one thingI realized using meteor: I like the way the toolchain works. Seems sensible to me in some points. Especially acknowledging, that meteor is a bit more than just npm/bower modules. I would be ok with maintaining that extra cost.

**Comment by robdodson** [comment]
Going to close this issue as there's not much actionable feedback. We've tried to simplify the process of managing a polymer app by moving more of the tooling into the Polymer CLI so developers don't have to wrangle their own gulp files. Hope that helps!
